### PR TITLE
Implement a message for user account suspension, as per MSC3823.

### PR DESCRIPTION
**Pull request by Yoric**
Users may have their account temporarily suspended from a homeserver,
e.g. due to ToS violations. This prevents them from sending new messages
but not from reading messages or redacting their own messages.

This patch displays an error message comparable to "some of your messages have not been sent".

See https://github.com/matrix-org/synapse/pull/12845 for more details.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Implement a message for user account suspension, as per MSC3823. ([\#8722](https://github.com/matrix-org/matrix-react-sdk/pull/8722)). Contributed by @Yoric.

### CODE REVIEW

**Comment by Yoric** [comment]
The test failure seems to be a timeout in unrelated code.

File: src/components/structures/RoomStatusBar.tsx
Comment (turt2live) [inline src/components/structures/RoomStatusBar.tsx:197]: instead of nesting, let's use the `?` operator:

`if (m.error?.errcode === 'whatever')`

File: src/components/structures/RoomStatusBar.tsx
Comment (turt2live) [inline src/components/structures/RoomStatusBar.tsx:248]: this can be a single line

File: src/components/structures/RoomStatusBar.tsx
Comment (turt2live) [inline src/components/structures/RoomStatusBar.tsx:253]: multiline like this needs brackets of some variety.

File: src/components/structures/RoomStatusBar.tsx
Comment (turt2live) [inline src/components/structures/RoomStatusBar.tsx:251]: this anchor needs `noreferrer` and `noopener`

File: src/components/structures/RoomStatusBar.tsx
Comment (turt2live) [inline src/components/structures/RoomStatusBar.tsx:195]: while we're here, all 3 of these variables need types.

**Review by turt2live — CHANGES_REQUESTED** [review]
generally lgtm - just concerns about the code style (some of which should be enforced by the linter, but empirically aren't). 

Please also include screenshots to make design review easier, as they will need to look at the copy.

**Comment by Yoric** [comment]
Matching screenshot:
Screenshot from 2022-06-02 07-53-31

**Comment by Yoric** [comment]
Also, test added.

**Comment by t3chguy** [comment]
@Yoric that screenshot looks rather broken

image

It is likely even worse with different scaling/font settings.

**Comment by Yoric** [comment]
> @Yoric that screenshot looks rather broken
> 
> image
> 
> It is likely even worse with different scaling/font settings.

Yes, I believe that this is what what the status bar currently looks like on 1000x660 (the default size for cypress), even without my patch.

**Comment by Yoric** [comment]
The cypress test passes locally but fails with the following error message in CI:

```
Error: Can't use 'tar -xzf' extract archive file: /home/runner/work/_actions/_temp_00e45d6c-6a29-4cbf-a707-3052823681c4/7b22363b-0b56-4817-8fec-0cd31a12eaad.tar.gz. return code: 2.
```

I don't have the authorization to relaunch tests.

**Comment by t3chguy** [comment]
Synapse develop is broken again
https://github.com/matrix-org/synapse/runs/6704161119?check_suite_focus=true has the same failure

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts:42]: Moving things up, but I have no clue what makes this questionable. `TEXT` is used ~6 times around the code, why shouldn't it be a constant?

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts:57]: > can we not just call a synapse API to suspend the user?

Not yet.

File: src/components/views/voip/CallView.tsx
Comment (Yoric) [reply src/components/views/voip/CallView.tsx:422]: Merge error, I guess.

Comment (Yoric) [reply src/components/views/voip/CallView.tsx:422]: Although it does look like [257aaf5e50a03834fe44ffd2fa31dd19b8ce5b1a](https://github.com/matrix-org/matrix-react-sdk/pull/8735/files)

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: it should be a constant, but we don't usually mix all caps and lowerCamelCase in our definition bodies.

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: Testing shouldn't be blocked by a frontend implementation, otherwise we'd get nothing done :p 

This sitting on a branch/here is fine while the rest of the feature is worked out, particularly considering this will be needing design review as well.

For the test here it's moderately important that we don't just intercept random endpoints and instead use the API provided by Synapse to do things: this means getting an admin API to suspend a user.

Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: > Testing shouldn't be blocked by a frontend implementation, otherwise we'd get nothing done :p

I actually meant user-testing. That can't happen without a frontend.

> For the test here it's moderately important that we don't just intercept random endpoints and instead use the API provided by Synapse to do things: this means getting an admin API to suspend a user.

I don't understand why it's important, but I'll take your word on it.

However, designing an API and getting it landed in Synapse before we can do any kind of user test sounds like something that will take us back several weeks, at best. The entire idea of prototyping it as a spam-check module is to be able to iterate quickly and without friction. I would very much prefer that.

Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: hmm, I'm not aware of a user testing plan for this to be honest 😬 

Because we're doing these tests at the end-to-end layer, we should be ensuring that it's, well, end-to-end 😇 For endpoints and such, it means performing actions as "server admins" to suspend/deactivate users and reacting to how the real server will tell us about it. If we just intercept the `/send` endpoint then we're no better off than a unit test, which doesn't help us test the application state.

A spam checking module is a good option, however suspension would also need to interact in a lot of other ways to prevent sustained abuse towards the server.

**Comment by t3chguy** [comment]
> Here's a screenshot with a more reasonable 1280x1024:

Until you open the right panel, surely?

**Comment by Yoric** [comment]
> > Here's a screenshot with a more reasonable 1280x1024:
> 
> Until you open the right panel, surely?

You are right.

Note that the problem already shows up prior to this patch, though.

Screenshot from 2022-06-06 09-47-17

I would suggest fixing this for good in a followup issue, as I clearly do not have a fine enough understanding of the CSS of Element Web for this purpose, plus this feels like a bit too much scope extension for this PR.

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: > A spam checking module is a good option, however suspension would also need to interact in a lot of other ways to prevent sustained abuse towards the server.

Yes. The idea is to progressively experiment with this.

**Comment by Yoric** [comment]
Note: Latest cypress error seems to be a timeout in unrelated code.

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (turt2live) [inline cypress/integration/9-room-status-bar/account-suspension.ts:63]: we don't use this syntax:
```suggestion
        cy.wait(1000);
```

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (turt2live) [inline cypress/integration/9-room-status-bar/account-suspension.ts:103]: sorry, I really think this needs to be an admin API call and not something we intercept. We should be testing that we behave correctly when Synapse tells us the account is suspended, not when we lie to it (otherwise we'd write an entirely different, less helpful, test).

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (turt2live) [inline cypress/integration/9-room-status-bar/account-suspension.ts:109]: this seems unrelated to the function - it needs comments, or moving, or both.

File: src/i18n/strings/en_EN.json
Comment (turt2live) [inline src/i18n/strings/en_EN.json:3095]: as already mentioned by @t3chguy, the status bar itself looks broken with this change, so we'd need to fix that before landing this PR.

The options for that are:
1. Wait for someone on the web team to get to it in the backlog (slow, and likely way too slow for your timelines).
2. Get your hands dirty and learn some CSS 😇 
3. Find someone who is able to help with writing that CSS and have them contribute ideas/guidance/code to this PR

Personally, I'd encourage option 2, particularly if you're expecting to be in the Element Web area for future changes.

Comment (turt2live) [inline src/i18n/strings/en_EN.json:3095]: for a future review iteration: design will need to be consulted for the message copy here.

**Review by turt2live — CHANGES_REQUESTED** [review]

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts:63]: Will do, but may I suggest amending this policy? 16_777_384 is much more readable than 16777384.

Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts:63]: I personally disagree on that, tbh :p 

I'm not sure what others on the team think, but I suspect since we haven't done so up until now we aren't likely to remember to do it, so will end up fighting the linter instead of working with it.

File: src/i18n/strings/en_EN.json
Comment (Yoric) [reply src/i18n/strings/en_EN.json:3095]: As already responded to @t3chguy, the status bar itself is already broken prior to this change.

I would really, really, really like to avoid scope creep. Can we please postpone this to a followup issue?

Comment (Yoric) [reply src/i18n/strings/en_EN.json:3095]: Who exactly?

Comment (turt2live) [reply src/i18n/strings/en_EN.json:3095]: can either get copy review internally from Write Club, or we can request design review once this hits a later stage of code review.

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts:103]: That feels like it would block an ongoing project.

Could we proceed as follows?

1. Land it with the intercept.
2. Perform the tests that we cannot perform until this feature has landed.
3. Design an admin API.
4. Spend a few weeks (or however long is required) bikeshedding, implementing and landing the admin API(s) with back-end.
5. Once we have an admin API, add a second test (or replace the existing test with a new one) that uses the admin API?

File: src/i18n/strings/en_EN.json
Comment (Yoric) [reply src/i18n/strings/en_EN.json:3095]: Who's/what's Write Club?

Comment (turt2live) [reply src/i18n/strings/en_EN.json:3095]: Even if it was broken before, we can't reasonably land something which has a near-100% chance of causing the bug. It just doesn't line up with our code quality requirements, sorry.

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts:103]: It would block the project, yes. Landing partial or incomplete code isn't viable as an option: we've literally never returned to temporary code, despite several promises that we will. We no longer take the chance, sorry.

File: src/i18n/strings/en_EN.json
Comment (turt2live) [reply src/i18n/strings/en_EN.json:3095]: Search the corp space :)

File: cypress/integration/9-room-status-bar/account-suspension.ts
Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: @turt2live Would you be more comfortable with the idea of letting this go through if we made the following changes?

1. Open a MSC for an error code (and associated field) `M_BLOCKED_BY_SPAMCHECK` (plus a prefixed version), dedicated specifically to being sent by spam-checker.
2. Rewrite all the user testing code we have to issue `M_BLOCKED_BY_SPAMCHECK` (prefixed).
3. Rewrite the current PR to use `M_BLOCKED_BY_SPAMCHECK` (prefixed).

This would avoid the current conundrum of pretending-to-be-a-synapse-feature because it clarifies that this is not a synapse feature but rather a spam-checker feature.

Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: I don't think an MSC would be needed for such an error code as that'd be a bit too far on the implementation-specific side of things.

We would still need tests that prove the spam checker error code is showing up appropriately, which is still non-zero levels of effort.

Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: > We would still need tests that prove the spam checker error code is showing up appropriately, which is still non-zero levels of effort.

It feels to me that this is the responsibility of the spam-checker, rather than that of matrix-react-sdk. Do you concur?

Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: It's not a challenge to detect the error code, but we absolutely need to have the `if` statement tested because otherwise it will break. Our end-to-end tests are deliberately set up to *not* mock error responses as we feel it's important to test that the actual server implementation works. 

If we were interested in mocked errors, we'd have used Complement or similar to set up a mocked homeserver. We picked Synapse to ensure we're able to function with a proper server, though. 

We cannot land this without suitable tests, so to unblock this the tests need to be written. It is currently the only blocker for this PR.

Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: > We cannot land this without suitable tests, so to unblock this the tests need to be written. It is currently the only blocker for this PR.

I'm currently doing my best to find a way to write a test that matches your criteria without having to:

1. throw away several weeks worth of work by three people;
2. start several more weeks of work in a different direction that doesn't match our requirements.

Is there any chance you could find an alternative solution?

Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: > A plausible test is one which installs the spam checker module onto a dedicated synapse (since we can support running different flavours of Synapse already) and trigger that with a message of some kind.

Very well, what's the procedure to create a dedicated Synapse for use with cypress?

Comment (turt2live) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: We have a "consent" template already that I think you should be able to largely copy/paste. Visiting [#element-dev:matrix.org](https://matrix.to/#/#element-dev:matrix.org) would be better than asking here though, to spread the support load and get an answer more quickly.

Comment (Yoric) [reply cypress/integration/9-room-status-bar/account-suspension.ts]: Thank you.

**Comment by andybalaam** [comment]
It looks like progress here has halted for now. Thank you for your work on this. The code and the Netlify build will continue to exist, so this can still be used to validate the MSC if needed, but I'm going to close the PR until it's being actively worked on.

**Pull request closed without merging**
